### ISSUE TITLE
noIntersperse never parses options

### ISSUE THREAD

**Issue by Thhethssmuz** [pcapriotti/optparse-applicative#233]
I am trying to create a parser for a program that is a runner for another program, and I also want to pass arguments along to this other program. Ideally I want to do this without having to use the `--` option after the main `file` argument since I know that all options after this argument should just be passed along.

From what I can gather, from #22, this seams to be the intended use case for the `noIntersperse` mode. And from them Haddocks:

> `noIntersperse :: InfoMod a`
> 
> Disable parsing of regular options after arguments

I take this to mean that regular options should be parsed prior to any argument. However, it seams that this option does not parse options at all:

``` haskell
data Options = Options
  { foo  :: Bool
  , file :: String
  , args :: [String]
  } deriving Show

parser :: Parser Options
parser = Options
  <$> switch ( long "foo" )
  <*> strArgument ( metavar "FILE" )
  <*> many ( strArgument ( metavar "ARGS..." ) )

main :: IO ()
main = do
  opts <- execParser $ info (helper <*> parser) noIntersperse
  print opts
```

When I run the above code with `--foo myfile -a -b -c` I expect to get:

```
Options {foo = True, file = "myfile", args = ["-a","-b","-c"]}
```

but instead I get:

```
Options {foo = False, file = "--foo", args = ["myfile","-a","-b","-c"]}
```

It also breaks the `help` option:

```
Options {foo = False, file = "--help", args = []}
```

Am I missing something here?

**Comment by HuwCampbell** [comment]
No you're not, I think this counts as a bug.

Currently turning on noIntersperse has the same behaviour as going after `--`, I'll have a think about a solution for you, and will add your test to the test suite.

**Comment by HuwCampbell** [comment]
I haven't finished this yet, but please see #234

**Comment by Thhethssmuz** [comment]
Ok, I've tried your branch and it seams to work for most cases. Though, and this might just be me trying to do something weird, but if you combine `noIntersperse` with a command parser:

```haskell
data Options = Options
  { foo  :: Bool
  , file :: String
  , args :: [String]
  } deriving Show

cmd :: ParserInfo Bool
cmd = info ( switch $ long "foo" ) idm

parser :: Parser Options
parser = Options
  <$> subparser ( command "cmd" cmd )
  <*> strArgument ( metavar "FILE" )
  <*> many ( strArgument $ metavar "ARGS..." )

main :: IO ()
main = do
  opts <- execParser $ info (helper <*> parser) noIntersperse
  print opts
```

And give it the input `cmd --bar myfile -a -b -c` I'd imagine you would get an error of type `Invalid option '--bar'`, but instead you get:

    Options {foo = False, file = "--bar", args = ["myfile","-a","-b","-c"]}

I guess this might be kind of an edge case as it is not necessarily well defined whether or not commands should trigger `noIntersperse` mode or not in the same way as arguments. And in this case it would be trivial to move the file and the forwarded options into the subparser and turn on `noIntersperse` there. However, if you had multiple subcommands it would be really practical to be able to use backtracking in this way.

I don't know if you agree with that or not or if this is just one of those cases you haven't finished yet... just kind of wanted to put it out there.

**Comment by HuwCampbell** [comment]
Indeed

**Comment by HuwCampbell** [comment]
Hmm, actually I think this might be the right behaviour.

If `--bar` was supposed to be used as an option, it should have gone before first positional, `cmd`.

**Comment by HuwCampbell** [comment]
Closed with #234
